File: week07/snippets/trim_newlines.mjs
import fs from 'fs/promises';

const filename = '/home/mjumbewu/Code/musa/musa-509/spring-2023/solutions/assignment02/data/septa_rail/stops copy.txt';
const f = await fs.open(filename, 'r+');

// Find the length of the file in bytes
const stats = await f.stat();

// Check whether the last byte is a newline (\n or \r)
let readPos = stats.size - 1;
let lastByte = await f.read(Buffer.alloc(1), 0, 1, readPos);
while ([10, 13].includes(lastByte.buffer[0])) {

  // If it is, trim it off
  console.log(`Trimming ${lastByte.buffer[0]} from end of file`);
  await f.truncate(readPos);

  // Go back one more byte and check again
  readPos--;
  lastByte = await f.read(Buffer.alloc(1), 0, 1, readPos);
}

File: week07/snippets/trim_newlines.py


filename = '...'
with open(filename, 'rb+') as f:

    # Go to the end of the file, and back up one byte
    f.seek(-1, 2)

    # Check whether the last byte is a newline (\n or \r)
    last_byte = f.peek(1)
    while last_byte in (b'\n', b'\r'):

        # If it is, trim it off
        print(f'Trimming {last_byte!r} from end of file')
        f.truncate()

        # Go back one more byte and check again
        f.seek(-1, 1)
        last_byte = f.peek(1)
